RecoParticleFlow/PFSimProducer/plugins/HGCalTrackCollectionProducer.cc

```diff
@@ -0,0 +1,190 @@
+#include "DataFormats/TrackReco/interface/Track.h"
+
+
+// for track propagation through HGC  
+// N.B. we are only propogating to first layer, so check these later
+#include "DataFormats/TrajectorySeed/interface/PropagationDirection.h"
+#include "TrackingTools/TrajectoryState/interface/TrajectoryStateTransform.h"
+#include "TrackingTools/MaterialEffects/interface/PropagatorWithMaterial.h"
+#include "DataFormats/GeometrySurface/interface/BoundDisk.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "MagneticField/Engine/interface/MagneticField.h"
+#include "Geometry/HGCalGeometry/interface/HGCalGeometry.h"
+
+//geometry records
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "MagneticField/Records/interface/IdealMagneticFieldRecord.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include<unordered_map>
+
+class HGCalTrackCollectionProducer : public edm::stream::EDProducer<> {
+
+public:
+  HGCalTrackCollectionProducer( const edm::ParameterSet & );
+private:
+  void produce( edm::Event &, const edm::EventSetup & ) override;
+  void beginLuminosityBlock(const edm::LuminosityBlock&, 
+			    const edm::EventSetup&) override;
+
+  edm::EDGetTokenT<edm::View<reco::PFRecTrack> > src_;
+  
+  // variables needed for copied goodPtResolution function
+  // need to go back and figure out sensible values
+  bool debug_;
+  const std::vector<double> DPtovPtCut_;
+  const std::vector<unsigned> NHitCut_;
+  const bool useIterTracking_;
+  //  const bool _useFirstLayerOnly; // always true now
+
+  // variables needed for copied extrapolation
+  edm::ESHandle<MagneticField> bField_;
+  edm::ESHandle<TrackerGeometry> tkGeom_;
+  std::array<std::string,1> hgc_names_; // 3 --> 1; extrapolate to hgcee only
+  std::array<edm::ESHandle<HGCalGeometry>,1> hgcGeometries_; // 3 --> 1; extrapolate to hgcee only
+  std::array<std::vector<ReferenceCountingPointer<BoundDisk> >,1> plusSurface_,minusSurface_; // 3 --> 1; extrapolate to hgcee only
+  std::unique_ptr<PropagatorWithMaterial> mat_prop_;
+
+  float diskOuterRadius_;
+  float diskInnerRadius_;
+};
+
+HGCalTrackCollectionProducer::HGCalTrackCollectionProducer(const edm::ParameterSet & iConfig) :
+  src_(consumes<edm::View<reco::PFRecTrack> >(iConfig.getParameter<edm::InputTag> ("src"))),
+  debug_(iConfig.getParameter<bool>("debug")),
+  DPtovPtCut_(iConfig.getParameter<std::vector<double> >("DPtOverPtCuts_byTrackAlgo")),
+  NHitCut_(iConfig.getParameter<std::vector<unsigned> >("NHitCuts_byTrackAlgo")),
+  useIterTracking_(iConfig.getParameter<bool>("useIterativeTracking"))
+  //  _useFirstLayerOnly(iConfig.getParameter<bool>("UseFirstLayerOnly")) // always true now
+{
+
+  if (debug_) std::cout << " HGCalTrackCollectionProducer::HGCalTrackCollectionProducer " << std::endl;
```

Comment:
Note that "cout" is not thread safe and will produce garbled output if the code is running multithreaded.
Better change to "LogInfo" (not printed by MessageLogger by default) or "LogPrint" (printed by MessageLogger by default) to use thread safe version.
We typically allow the pattern ``` if (debug) cout << ... ```, but it should be understood that it has issues and will not allow debugging of a job running with multiple threads.